app/models/project_services/composer_service.rb

```diff
@@ -0,0 +1,459 @@
+        ]
+      },
+      { type: 'fieldset', legend: 'Tags:', fields:
+        [
+          { type: 'checkbox', name: 'export_tags', label: 'Tag Listing' },
+          { type: 'text', name: 'tag_filters', label: 'Tag Filters', placeholder: 'tags you wish to export comma seperated.', hint: 'Separate tags with commas. Leave blank to export all tags.' }
+        ]
+      },
+      { type: 'fieldset', legend: 'Advanced:', fields:
+        [
+          { type: 'textarea', name: 'custom_json', label: 'Custom JSON', placeholder: 'custom composer.json to use for exporting this package.'},
+        ]
+      }
+    ]
+  end
+
+  #disable test button
+  def can_test?
+    false
+  end
+
+  def process_packages_on_save
+
+      #process packages for all tags
+      project.repository.tags.each do |tag|
+        process_commit(tag)
+      end
+
+      # process packages for all branches
+      project.repository.branches.each do |branch|
+        process_commit(branch)
+      end
+
+  end
+
+  def process_commit(ref)
+
+    previous_package_removed = false
+    begin
+      if commit_was_activated? && commit_was_exported?(ref)
+        defaults = package_mode_was == 'advanced' ? ActiveSupport::JSON.decode(custom_json_was) : {"type"=>package_type_was}
+        package = Composer::Package.new(project, ref, package_mode_was, defaults)
+        manager.rm_package(package)
+        previous_package_removed = true
+      end
+    rescue Exception => e
+      # Skip on error
+      log("remove previous ref exception #{e}")
+    end
+
+    begin
+      defaults = package_mode == 'advanced' ? ActiveSupport::JSON.decode(custom_json) : {"type"=>package_type}
+      package = Composer::Package.new(project, ref, package_mode, defaults)
+      if activated? && commit_is_exported?(ref)
+        manager.add_package(package)
+      elsif not previous_package_removed
+        manager.rm_package(package)
+      end
+    rescue Exception => e
+      # Skip on error
+      action = (activated? && commit_is_exported?(ref)) ? "add" : "remove"
+      log("#{action} ref exception #{e}")
+    end
+
+  end
+
+  def execute(push_data)
+
+    oldrev = push_data[:before]
+    newrev = push_data[:after]
+    ref = push_data[:ref]
+
+    # sync our changes
+    if newrev == Gitlab::Git::BLANK_SHA # push delete
+
+      # recreate exported packages since we dont have access to the original push
+      manager.clear_packages
+
+      #process packages for all tags
+      project.repository.tags.each do |tag|
+        process_commit(tag)
+      end
+
+      # process packages for all branches
+      project.repository.branches.each do |branch|
+        process_commit(branch)
+      end
+
+    else # push create / modify
+
+      if push_to_branch?(ref)
+        if (branch = project.repository.branches.find { |branch| branch.name == branch_name(ref) && branch.target == newrev })
```

Comment:
Shadowing outer local variable - `branch`.<br>Line is too long. [126/80]<br>Prefer `detect` over `find`.